pkg/kkrtpsi/sender.go

```diff
@@ -0,0 +1,282 @@
+		oprfInputSize = int(cuckoo.Factor * float64(remoteN))
+		if 1 > oprfInputSize {
+			oprfInputSize = 1
+		}
+
+		// instantiate an AES block
+		aesBlock, err := aes.NewCipher(secretKey)
+		if err != nil {
+			return err
+		}
+
+		// exhaust local ids, and precompute all potential
+		// hashes and store them using the same
+		// cuckoo hash table parameters as the receiver.
+		go func() {
+			cuckooHasher := cuckoo.NewCuckooHasher(uint64(remoteN), seeds)
+
+			// prepare struct to send inputs and hasher to stage 3
+			var result stage1Result
+			result.inputs = make([]inputToOprfEncode, n)
+
+			var i int
+			for id := range identifiers {
+				// hash and calculate pseudorandom code given each possible hash index
+				var bytes [cuckoo.Nhash][]byte
+				for hIdx := 0; hIdx < cuckoo.Nhash; hIdx++ {
+					bytes[hIdx] = crypto.PseudorandomCode(aesBlock, id, byte(hIdx))
+				}
+				result.inputs[i] = inputToOprfEncode{prcEncoded: bytes, bucketIdx: cuckooHasher.BucketIndices(id)}
+				i++
+			}
+
+			result.hasher = cuckooHasher.GetHasher()
+			encodedInputChan <- result
+		}()
+
+		// end stage1
+		timer, mem = printStageStats(logger, 1, start, start, 0)
+		logger.V(1).Info("Finished stage 1")
+		return nil
+	}
+
+	// stage 2: act as sender in OPRF, and receive OPRF keys
+	stage2 := func() error {
+		logger.V(1).Info("Starting stage 2")
+
+		// instantiate OPRF sender with agreed parameters
+		oprfKey, err = oprf.NewOPRF(oprfInputSize).Send(s.rw)
+		if err != nil {
+			return err
+		}
+
+		// end stage2
+		timer, mem = printStageStats(logger, 2, timer, start, mem)
+		logger.V(1).Info("Finished stage 2")
+		return nil
+	}
+
+	// stage 3: compute all possible OPRF output using keys obtained from stage2
+	stage3 := func() error {
+		logger.V(1).Info("Starting stage 3")
+
+		// inform the receiver the number of local ID
+		if err := binary.Write(s.rw, binary.BigEndian, &n); err != nil {
+			return err
+		}
+
+		message := <-encodedInputChan
+		nWorkers := runtime.GOMAXPROCS(0)
+		var localEncodings = make(chan [][cuckoo.Nhash]uint64, nWorkers*2)
+
+		batchSize := 2048
+		nBatches := len(message.inputs) / batchSize
+		workerResp := nBatches / nWorkers
+
+		g, ctx := errgroup.WithContext(ctx)
+
+		// each worker is responsible to encode and hash workerResp batches and send them out
+		for w := 0; w < nWorkers; w++ {
+			w := w
+			for batchNumber := 0; batchNumber < workerResp; batchNumber++ {
+				batchNumber := batchNumber
+				// each batch is done by a goroutine
+				g.Go(func() error {
+					batch := make([][cuckoo.Nhash]uint64, batchSize)
+					for bIdx := 0; bIdx < batchSize; bIdx++ {
+						batch[bIdx] = message.inputs[(w*workerResp+batchNumber)*batchSize+bIdx].encodeAndHash(oprfKey, message.hasher)
+					}
+
+					// batch is filled; send it out
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					case localEncodings <- batch:
+						return nil
+					}
+				})
+			}
+		}
+
+		// extra worker deals with the remaining inputs
```

Comment:
Specify that we devolve to single thread when nBatches < nWorkers since the above workers will exit early.